<a href="https://colab.research.google.com/github/mhdhaikal22/data-science-2026/blob/main/Pertemuan10_%5BMUHAMMAD_HAIKAL%5D_%5B240401010269%5D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Pertemuan 10 - Algoritma Klasifikasi (Bagian 2)

**Nama:** Muhammad Haikal  
**NIM:** 240401010269  
**Kelas:** IF401

In [ ]:
# Muat dan Eksplorasi Data

import pandas as pd

from google.colab import drive
drive.mount('/content/drive')

df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/WA_Fn-UseC_-Telco-Customer-Churn.csv")

print(df.shape)
print(df["Churn"].value_counts(normalize=True))

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
(7043, 21)
Churn
No     0.73463
Yes    0.26537
Name: proportion, dtype: float64


In [ ]:
# Preprocessing

from sklearn.model_selection import train_test_split
import pandas as pd

# Encoding fitur kategorikal
X = pd.get_dummies(df.drop("Churn", axis=1), drop_first=True)
y = df["Churn"]

# Data splitting
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    stratify=y,
    random_state=42
)

In [ ]:
# Latih Model

from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(
    n_estimators=300,
    class_weight="balanced",
    random_state=42
)

rf.fit(X_train, y_train)

RandomForestClassifier(class_weight='balanced', n_estimators=300,
                       random_state=42)

In [ ]:
# Evaluasi

from sklearn.metrics import classification_report, roc_auc_score

# Prediksi kelas
y_pred = rf.predict(X_test)

# Prediksi probabilitas (untuk ROC-AUC)
y_proba = rf.predict_proba(X_test)[:, 1]

# Classification report
print("=== Classification Report ===")
print(classification_report(y_test, y_pred))

# ROC-AUC score
roc_auc = roc_auc_score(y_test, y_proba)
print("ROC-AUC Score:", roc_auc)

=== Classification Report ===
              precision    recall  f1-score   support

          No       0.84      0.90      0.87      1035
         Yes       0.65      0.52      0.58       374

    accuracy                           0.80      1409
   macro avg       0.74      0.71      0.72      1409
weighted avg       0.79      0.80      0.79      1409

ROC-AUC Score: 0.8337260068717869


In [ ]:
# Prediksi Probabilitas dan Simpulkan

import pandas as pd

# Prediksi probabilitas churn
proba_churn = rf.predict_proba(X_test)[:, 1]

# Tampilkan beberapa contoh probabilitas
print("Probabilitas churn (10 data pertama):")
print(proba_churn[:10])

# Gabungkan hasil ke dalam DataFrame
hasil = pd.DataFrame({
    "Actual": y_test.values,
    "Predicted": y_pred,
    "Prob_Churn": proba_churn
})

hasil.head(10)

Probabilitas churn (10 data pertama):
[0.00333333 0.63666667 0.07333333 0.37666667 0.01666667 0.56666667
 0.46333333 0.03       0.00333333 0.53      ]


,Actual,Predicted,Prob_Churn
0,No,No,0.003333
1,No,Yes,0.636667
2,No,No,0.073333
3,No,No,0.376667
4,No,No,0.016667
5,No,Yes,0.566667
6,No,No,0.463333
7,No,No,0.030000
8,No,No,0.003333
9,Yes,Yes,0.530000


Model Random Forest yang dibangun mampu memprediksi probabilitas churn pelanggan menggunakan fungsi predict_proba. Berdasarkan evaluasi, model memiliki performa yang cukup baik dengan ROC-AUC sekitar 0.83 yang menunjukkan kemampuan pemisahan kelas yang baik. Namun, recall pada kelas churn masih belum optimal sehingga beberapa pelanggan berisiko churn belum terdeteksi. Model masih dapat ditingkatkan dengan tuning parameter atau penyesuaian threshold prediksi.